Cubical/Homotopy/WhiteheadsLemma.agda

```diff
@@ -0,0 +1,447 @@
+
+  isoPreComp≃∙ : {A : Type ℓ} {a b c : A} {q : b ≡ c} (p : a ≡ b)
+    → ((b ≡ c) , q) ≃∙ ((a ≡ c) , p ∙ q)
+  fst (isoPreComp≃∙ p) = isoToEquiv (isoPreComp p)
+  snd (isoPreComp≃∙ p) = refl
+
+  PathPCongLemma : {A B : Type ℓ}
+    (f : A → B)
+    (a a' : A)
+    (r : a ≡ a')
+    (b : B)
+    (p : f a ≡ b)
+    (q : f a' ≡ b)
+    → cong f r ≡ p ∙ (sym q)
+    → PathP (λ i → f (r i) ≡ b) p q
+  PathPCongLemma {B = B} f a a' =
+    J (λ a'' r → (b : B) (p : f a ≡ b) (q : f a'' ≡ b)
+       → cong f r ≡ p ∙ (sym q)
+       → PathP (λ i → f (r i) ≡ b) p q)
+      (λ b p q r →
+       sym (lUnit q ∙ cong (_∙ q) r ∙ sym (assoc p (sym q) q)
+                    ∙ cong (p ∙_) (lCancel q) ∙ sym (rUnit p)))
+
+  module _ (F : Pointed ℓ → Pointed ℓ)
+           (F' : {A B : Pointed ℓ} → (A →∙ B) → (F A →∙ F B)) where
+
+    specialCong∙ : {A B C D : Pointed ℓ} (p : A ≡ C) (q : B ≡ D)
+                   (f : A →∙ B) (g : C →∙ D)
+                   (r : PathP (λ i → (p i) →∙ (q i)) f g)
+                → PathP (λ i → (cong F p) i →∙ (cong F q) i) (F' f) (F' g)
+    specialCong∙ {A = A} {B = B} {C = C} {D = D} =
+      J ( λ C' p →
+                    (q : B ≡ D) (f : A →∙ B) (g : C' →∙ D)
+                    (r : PathP (λ i → (p i) →∙ (q i)) f g)
+                 → PathP (λ i → (cong F p) i →∙ (cong F q) i) (F' f) (F' g))
+        ( J ( λ D' q →
+                        (f : A →∙ B) (g : A →∙ D')
+                        (r : PathP (λ i → A →∙ (q i)) f g)
+                     → PathP (λ i → F A →∙ (cong F q) i) (F' f) (F' g))
+            ( λ f g r → cong F' r))
+
+  ∙PathP→Square : {A B C D : Pointed ℓ} (p : A ≡ C) (q : B ≡ D)
+    (f : A →∙ B) (g : C →∙ D)
+    → PathP (λ i → (p i) →∙ (q i)) f g
+    → (Iso.inv (pathToIso (λ i → fst (q i))))
+     ∘ (fst g)
+     ∘ (Iso.fun (pathToIso (λ i → fst (p i))))
+     ≡ (fst f)
+  ∙PathP→Square {A = A} {B = B} {C = C} {D = D} =
+    J ( λ C' p →
+                  (q : B ≡ D) (f : A →∙ B) (g : C' →∙ D)
+               → PathP (λ i → (p i) →∙ (q i)) f g
+               → (Iso.inv (pathToIso (λ i → fst (q i))))
+                 ∘ (fst g)
+                 ∘ (Iso.fun (pathToIso (λ i → fst (p i))))
+                 ≡ (fst f))
+      ( J ( λ D' q →
+                      (f : A →∙ B) (g : A →∙ D')
+                   → PathP (λ i → A →∙ (q i)) f g
+                   → (Iso.inv (pathToIso (λ i → fst (q i))))
+                    ∘ (fst g)
+                    ∘ (Iso.fun (pathToIso (λ i → (fst A))))
+                    ≡ (fst f))
+          ( λ f g r →
+            funExt ( λ a → transportRefl (fst g (transport refl a))
+                          ∙ cong (fst g) (transportRefl a)
+                          ∙ λ i → (fst (r (~ i))) a)))
+
+  doubleCompPath-filler-refl : {A : Type ℓ} {a b : A} (p : a ≡ b)
+    → p ≡ refl ∙∙ p ∙∙ refl
+  doubleCompPath-filler-refl p = doubleCompPath-filler refl p refl
+
+  isContr→isContr→isEquiv : {A B : Type ℓ} (f : A → B)
```

Comment:
`isContr→Iso` in Foundations.Isomorphism

Reply:
More precisely `isEquivFromIsContr` in Foundations.Equiv.Properties